syntax = "proto3";

package dynamic_density;

import "google/protobuf/timestamp.proto";

service DynamicDensityService {
  rpc RPCQuery(RPCQueryParams) returns (RPCQueryResult) {}
}

message RPCQueryParams {
  oneof request {
    ListDensityMapsParams list_density_maps_params = 1;
    DensityMapIdentifier get_map_with_identifier = 2;
    DensityMapDescription set_density_map_description = 3;
  }
}

message RPCQueryResult {
  oneof result {
    ListDensityMapsResult list_density_maps_result = 1;
    DensityMap density_map_result = 2;
  }
}

message DensityMapIdentifier {
  int32 identity = 1;
}

message DensityMapDescription {
  enum Type {
    HISTOGRAM = 0;
    KDE = 1;
    KDE2D = 2;
  }

  DensityMapIdentifier identifier = 1;
  string title = 2;
  repeated string labels = 3;
  google.protobuf.Timestamp timestamp = 4;
  double decay_rate = 5;
  Type type = 6;
}

message ListDensityMapsParams {}

message ListDensityMapsResult {
  repeated DensityMapDescription descriptions = 1;
}

message DynamicHistogram {
  DensityMapDescription description = 1;
  repeated double bounds = 2;
  repeated double counts = 3;
}

message DynamicKDE {
  // Kernel represents 1-d data by using a single value for both coord and
  // variance and no covariance. It represents 2-d data with two values for
  // each coord and variance and a single value for covariance.
  message Kernel {
    repeated double coord = 1;
    repeated double variance = 2;
    double covariance = 3;
    double count = 4;
  }

  DensityMapDescription description = 1;
  repeated Kernel kernels = 2;
}

message DensityMap {
    oneof density_map {
        DynamicHistogram dynamic_histogram = 1;
        DynamicKDE dynamic_kde = 2;
    }
}
